### ISSUE TITLE
Emotion 11

### ISSUE THREAD

**Issue by mitchellhamilton** [emotion-js/emotion#1606]
We're planning to do a major version of Emotion soon, we don't have a definite timeline but it should be out before the end of the year, most likely earlier.

We don't currently have any massive API changes planned but there will be some package renaming which should be possible to migrate with codemods alone. (This isn't anything like the v9 -> v10 change)

The big changes that we are currently doing:
- Package renaming (#1635)
- Use hooks internally which means a better tree in React DevTools and a smaller bundle size (#967)
- Rewrite TypeScript definitions (#1501)
- Allow mapping of imports in the babel plugin (#1220) 

We'll update this issue for larger planned changes. For smaller changes, please see the [v11 milestone](https://github.com/emotion-js/emotion/milestone/1). The issues on the milestone aren't necessarily going to be addressed in v11 but we'll consider them.

We are already doing pre-releases which will be available on the `next` dist tag on npm.

**Comment by kevinrodriguez-io** [comment]
Regarding to @emotion/native@next

For some reason I'm not getting the types folder installed on my node_modules and all my types are blank.

Is there any reason for this?

**Comment by Andarist** [comment]
The version with those typings has not been published yet, I will do that in a moment.

**Comment by kevinrodriguez-io** [comment]
@Andarist Thanks a lot! The new typings are working great.

I'm getting theme intellisense by doing this:

types.ts
```ts
interface Theme {
  colors: {
    // Color palette here
  },
  fonts: {
    sansSerif: {
      thin: string,
      regular: string,
      bold: string,
      black: string
    }
  }
}

interface EmotionComponentProps {
  theme?: Theme
}
```

PinDot.ts
```ts
import styled from '@emotion/native'
import { EmotionComponentProps } from 'types'

interface PinDotProps extends EmotionComponentProps {
  active: boolean
}

const PinDot = styled.View<PinDotProps>(
  {
    width: 16,
    height: 16,
    borderRadius: 16,
  },
  ({ active, theme }) => ({
    backgroundColor: active
      ? theme.colors.flatPurple.light
      : theme.colors.flatWhite.dark,
  }),
)

export default PinDot
```

**Comment by thewavelet** [comment]
Thank you for your great awesome works! And I have two questions. 
When is v11 going to release?
Can I use v11 in advance with yarn add @emotion/core@next @emotion/native@next commands?
Because I need react native typings.

**Comment by kevinrodriguez-io** [comment]
@thewavelet You can add it by using:

`yarn add @emotion/react@next @emotion/native@next`

You'll have to remove @emotion/core and emotion-theming, both packages are now part of @emotion/react

**Comment by Andarist** [comment]
> When is v11 going to release?

When it's ready 😉

**Comment by thewavelet** [comment]
Thank you! I will use emotion instead of styled-components.

**Comment by hammadj** [comment]
I'm having trouble using `@emotion/native` this in Expo Web / react-native-web. I get the following error when compiling the web target:
```
 web  Starting Webpack on port 19006 in development mode.
  build [====================] 100%

 web  Failed to compile.
.../rn-app/node_modules/@react-navigation/bottom-tabs/lib/module/views/BottomTabBar.js
Module not found: Can't resolve 'react-native-web/dist/exports/Animated' in '.../rn-app/node_modules/@react-navigation/bottom-tabs/lib/module/views'
```

Any guidance here?

**Comment by hammadj** [comment]
Is there plans to improve support for the css prop in `@emotion/native` ? Seems to be the new recommended approach instead of using the `styled.View` syntax.

**Comment by NoobsArePeople2** [comment]
> > When is v11 going to release?
> 
> When it's ready 😉

Hate to be the guy who piles on this question but I need to ask.

I'm looking at migrating a significant amount of code from Emotion 9 to Emotion 10 in the near future and would like to avoid having to then migrate to Emotion 11 after completing the first migration. It sounds like the Emotion 11 API will be very similar to 10 so the level of effort in migrating 9 --> 11 is similar to 9 --> 10.

Can you provide a ballpark estimate on delivery?

**Comment by Andarist** [comment]
It's hard to tell - previously I thought that we are going to ship it in December, but now I hope for February. No promises though, we are doing it in our free time and there is no much of it lately. 

V11 is pretty much done though and is published on npm, so I would encourage you to try it out - even if it's not yet marked with the "latest" tag. 

The biggest leftover for v11 (which is not quite mentioned anywhere) is a stylis.js rewrite (our underlying parser). It's happening, just slowly - we are working out the plugin API (parser itself is ready).

**Comment by hammadj** [comment]
@Andarist  thanks for the update and for all the hard work!

**Comment by kevinrodriguez-io** [comment]
Turns out you don't need to declare an interface and inherit it everywhere 😄 you can use Typescript's Declaration merging!

emotion.d.ts

```ts
import '@emotion/react'

declare module '@emotion/react' {
  export interface Theme {
    colors: {
      primary: string
      secondary: string
      light: string
      dark: string
    }
    fonts: {
      sansSerif: {
        thin: string
        regular: string
        bold: string
        black: string
      }
    }
  }
}
```

**Comment by jsejcksn** [comment]
[It looks like](https://unpkg.com/browse/@emotion/react@11.0.0-next.11/dist/react.umd.min.js) the `@emotion/react` umd module is still being assigned to `window.emotionCore`. It should be changed to `window.emotionReact`, yes?

**Comment by Andarist** [comment]
@jsejcksn Yes, just changed it: https://github.com/emotion-js/emotion/commit/9e833bacd890a8bf9daa1bd910eb83901fde1cc4 . Should be available with the next release.

**Comment by chancestrickland** [comment]
This appears to be a catch-all issue for everything v11, so asking here: is there a reason that `CreateStyled` is no longer a generic interface that accepts a custom theme type def? I modified the types locally with what I *think* should be the intended types and can submit a PR if it's helpful, but let me know if I missed something here.

```tsx
declare module '@emotion/styled' {
  // ...
  export interface BaseCreateStyled<CustomTheme extends Theme = {}> {
    <
      C extends React.ComponentType<React.ComponentProps<C>>,
      ForwardedProps extends keyof React.ComponentProps<
        C
      > = keyof React.ComponentProps<C>
    >(
      component: C,
      options: FilteringStyledOptions<PropsOf<C>, ForwardedProps>
    ): CreateStyledComponent<
      Pick<PropsOf<C>, ForwardedProps> & { theme?: CustomTheme }
    >;

    <C extends React.ComponentType<React.ComponentProps<C>>>(
      component: C,
      options?: StyledOptions<PropsOf<C>>
    ): CreateStyledComponent<PropsOf<C> & { theme?: CustomTheme }>;

    <
      Tag extends keyof JSX.IntrinsicElements,
      ForwardedProps extends keyof JSX.IntrinsicElements[Tag] = keyof JSX.IntrinsicElements[Tag]
    >(
      tag: Tag,
      options: FilteringStyledOptions<
        JSX.IntrinsicElements[Tag],
        ForwardedProps
      >
    ): CreateStyledComponent<
      { theme?: CustomTheme },
      Pick<JSX.IntrinsicElements[Tag], ForwardedProps>
    >;

    <Tag extends keyof JSX.IntrinsicElements>(
      tag: Tag,
      options?: StyledOptions<JSX.IntrinsicElements[Tag]>
    ): CreateStyledComponent<
      { theme?: CustomTheme },
      JSX.IntrinsicElements[Tag]
    >;
  }

  export type StyledTags<CustomTheme extends Theme = {}> = {
    [Tag in keyof JSX.IntrinsicElements]: CreateStyledComponent<
      { theme?: CustomTheme },
      JSX.IntrinsicElements[Tag]
    >;
  };

  export interface CreateStyled<CustomTheme extends Theme = {}>
    extends BaseCreateStyled<CustomTheme>,
      StyledTags<CustomTheme> {}

  declare const styled: CreateStyled;
  export default styled;
}
```

**Comment by Andarist** [comment]
@chancestrickland we've changed how you can configure Theme type to what should (I believe) be way more convenient than the previous solution, see https://deploy-preview-1600--emotion.netlify.app/docs/typescript#define-a-theme

**Comment by chancestrickland** [comment]
Ah beautiful, I agree that works even better. Thanks @Andarist!

**Comment by Andarist** [comment]
No problem, hope you enjoy Emotion 11 - we hope to ship it soon-ish (but I also thought it was going to be shipped in 2019 😅).

**Comment by andrewphillipo** [comment]
Just in case you want to generate the theme type automatically rather than having to declare it you can see my solution here: https://github.com/emotion-js/emotion/issues/1861#issuecomment-624044744

**Comment by chancestrickland** [comment]
@Andarist any thoughts on still allowing the generic workaround in v11? I agree that the new solution is nicer, but are there any drawbacks to using both techniques for a less breaking change? Might be useful for migrating existing projects.

**Comment by Andarist** [comment]
I'm afraid it's not feasible on our side - would involve some work to bring it back for a rather low return value. There are overall not many impactful breaking changes scheduled for v11 so I hope it won't be too hard to migrate easily. Migrating this particular thing should be a matter of adding a declaration file in which you augment our interface + some find&replace on import sources. Thanks to `importMap` option introduced in our Babel plugin the latter step can even be avoided to a quicker migration.

**Comment by lucksp** [comment]
> Turns out you don't need to declare an interface and inherit it everywhere 😄 you can use Typescript's Declaration merging!
> 

How does this work in V10?  I am unable to use the `@next` version in production at work and running into issues: https://codesandbox.io/s/emotion-typescript-definition-r085c?from-embed

**Comment by andrewphillipo** [comment]
You can look at the docs for V10, I think you are trying to use the mechanism new to V11 for theming in an old version: https://emotion.sh/docs/typescript#define-a-theme 

Essentially you are creating a new version of "styled" in your codebase that you must use instead of the non theme aware version.

**Comment by diondirza** [comment]
Hey, how do you use `css` prop in v11? I tried with the same setup as v10 by installing `@emotion/babel-preset-css-prop@next` yet it still yields 
`Can't resolve '@emotion/core' error`

**Comment by Andarist** [comment]
@diondirza could you prepare a repro case? Note that `@emotion/core` has been renamed to `@emotion/react` in v11

**Comment by diondirza** [comment]
@Andarist please ignore my issue, probably because yarn didn't update my package correctly. I tried the first time install the plugin without `@next` then I install again with `@next`. Seems yarn not correctly install the package at that time. I remove `yarn.lock` and did `yarn install` again now it is working fine.

**Comment by Arcus16** [comment]
Hello, i use `"@emotion/native": "^11.0.0-next.12",` in my react native project `0.62.2`.  I tried to extend component to add some additional styles, but these styles are ignored.

`import styled from '@emotion/native';
import { View, ViewStyle } from 'react-native';

const Flex = styled(View)<ViewStyle>((props) => ({
  flexDirection: props.flexDirection,
  alignItems: props.alignItems,
  justifyContent: props.justifyContent,
  marginTop: props.marginTop,
}));

export const Flex_CC = styled(Flex)({
  alignItems: 'center',
  justifyContent: 'center',
});

export default Flex;`

**Comment by osdiab** [comment]
Any update on when this is expected to land? excited to try it out at my company :)

**Comment by Slapbox** [comment]
Will there be any changes in Emotion 11 that significantly impact performance, for better or for worse?

**Comment by Andarist** [comment]
Not sure if "significantly" - but yes, Emotion 11 should be faster (and smaller I think). Don't focus on those numbers too much though, this really doesn't matter that much for your apps.

**Comment by Slapbox** [comment]
Mostly just a question of curiosity. I've enjoyed watching the performance of CSS in JS improve over the past few years. When the trend started out the performance hit was pretty substantial. Now the only time I can really think of that it matters is when a user changes themes on the fly.

**Comment by ioveracker** [comment]
> You can add it by using:
> 
> `yarn add @emotion/react@next @emotion/native@next`
> 
> You'll have to remove @emotion/core and emotion-theming, both packages are now part of @emotion/react

Is anything else needed to get up and running with `v11.0.0-next.15`? I'm getting an error at runtime in a basic React Native app.

```
error: Error: While trying to resolve module `stylis` from file `/Users/ioveracker/Source/ioveracker/demo-react-native-emotion/node_modules/@emotion/cache/dist/cache.browser.cjs.js`, the package `/Users/ioveracker/Source/ioveracker/demo-react-native-emotion/node_modules/stylis/package.json` was successfully found. However, this package itself specifies a `main` module field that could not be resolved (`/Users/ioveracker/Source/ioveracker/demo-react-native-emotion/node_modules/stylis/dist/stylis.cjs`.
```
[Repro case here](https://github.com/ioveracker/demo-react-native-emotion/tree/error-resolving-stylis).

(Apologies if this belongs on a top-level issue. I'd be happy to move it if requested.)

**Comment by efoken** [comment]
Had the same problem, for Expo just update your app.json:

```json
{
  "expo": {
    ...
    "packagerOpts": {
      "sourceExts": ["cjs"]
    }
}
```

Or adapt your Metro bundler config. BTW, there is also an open issue for that in Metro repo https://github.com/facebook/metro/issues/535

**Comment by louisgv** [comment]
To those who're using react-native with emotion@next, I was wondering if you are also using typescript as well? After looking at the typing for emotion native, I decided to create a concept PR for how to make the typing more concise/DRY: 

https://github.com/emotion-js/emotion/pull/2000

Would love to learn how we can improve this

**Comment by anilanar** [comment]
I can confirm that 50k LoC web project upgraded to Emotion v11 @userlike and QA'ed. No bugs so far.

**Comment by Andarist** [comment]
@anilanar wow, that's great to hear! 

In other news - we are going to ship the first RC this week (almost identical to the latest v11 release) and probably ship it as stable within a month.

**Comment by szagi3891** [comment]
@Andarist - Fantastic news :)

**Comment by Slmii** [comment]
> @chancestrickland we've changed how you can configure Theme type to what should (I believe) be way more convenient than the previous solution, see https://deploy-preview-1600--emotion.netlify.app/docs/typescript#define-a-theme

this is not working for me. below I defined my theme types.
```js
import '@emotion/react';

declare module '@emotion/react' {
	export interface DefaultTheme {
		colors: {
			main: {
				primary: string;
				primaryDark: string;
				primaryHover: string;
				primaryDisabled: string;
				primaryDisabledDark: string;
			};
		}
	}
}
```
in my `theme.ts` I then use the type `DefaultTheme`
```js
import { DefaultTheme } from '@emotion/react';

const theme: DefaultTheme = {
	colors: {
		main: {
			primary: '#3885E9',
			primaryDark: '#2C6ABA',
			primaryHover: '#2177E6',
			primaryDisabled: '#7DAFF1',
			primaryDisabledDark: '#4F93EC'
		},
	}
}
```
in my styling components it then says
![image](https://user-images.githubusercontent.com/9927849/95786735-74624a80-0cd8-11eb-8c33-aa760c4af2e3.png)

This the `Theme`
```js
export type StyledTags = {
  [Tag in keyof JSX.IntrinsicElements]: CreateStyledComponent<
    {
      theme?: Theme
      as?: React.ElementType
    },
    JSX.IntrinsicElements[Tag]
  >
}

export interface Theme {}

```

**Comment by anilanar** [comment]
@Slmii It's supposed to be called `Theme`, not `DefaultTheme`.

**Comment by Slmii** [comment]
@anilanar thanks for the swift reply 👍  it's working, must've overlooked that part!

**Comment by zerofront** [comment]
#1321 So v11 won't include hooks like useStyle ?

**Comment by Andarist** [comment]
No - there is no much need for it from our PoV. I'll think about adding support for it with a Babel transform - but it's not high on my priority list.

**Comment by bardak-dev** [comment]
**@emotion/styled/macro** doesn't work like before.
Theme types works perfect now, thx!

**Comment by Andarist** [comment]
> @emotion/styled/macro doesn't work like before.

What do you mean?

**Comment by bardak-dev** [comment]
> > @emotion/styled/macro doesn't work like before.
> 
> What do you mean?

My bad, issue was in my code, in @emotion 11 everything works great!

**Comment by SimenB** [comment]
🎉

**Comment by eddyzera** [comment]
I have a problem when I use chakra ui along with emotion:

Error: The `@emotion/core` package has been renamed to `@emotion/react`. Please import it like this `import { jsx } from '@emotion/react'`.

**Comment by Andarist** [comment]
Please always try to share a repro case in a runnable form - either by providing a git repository to clone or a codesandbox. OSS maintainers usually can't afford the time to set up the repro, even if exact steps are given.

**Comment by ljosberinn** [comment]
@eddyzera you seem to be using either chakra v0 or unnecessarily added `@emotion/core` to a v1 project (currently, in v1, you do not need to install it separately). 
- if youre using v0, you'll have to pin emotion dependencies to 10.x, I believe `/core` to 10.0.27 or lower and `/styled` to 10.0.35 or lower
- if youre using v1, simply remove all emotion dependencies from your package json, remove lock file, remove node_modules and reinstall

v1 will most likely add `@emotion/react` as well as a few others as peer dependencies, but for the time being this will fix it.

**Comment by Fabrus98** [comment]
@ljosberinn hi, i just got the error
`./node_modules/@chakra-ui/core/dist/es/Accordion/index.js
Module not found: Can't resolve '@emotion/core' in 'C:\Users\Administrateur\Desktop\ItalianHUBot\dashboard-client\node_modules\@chakra-ui\core\dist\es\Accordion'`
trying to use @chakra-ui. I don't know how to fix it, i'm learning use react.js.
Hope someone could help me, thanks.

**Comment by Andarist** [comment]
Either install `@emotion/core` or upgrade your Chakra version to v1

**Comment by ljosberinn** [comment]
@Fabrus98 please file an issue in the chakra-ui repository, not here. alternatively, join our Chakra discord for faster triaging: https://discord.gg/vw7tQnvp

**Comment by Vadorequest** [comment]
I'm a bit lost by reading https://emotion.sh/docs/emotion-11, I'm migrating from v10 to v11, using Next.js.

Here is a simple component:

```jsx
import { css } from '@emotion/react';
import React from 'react';

import AnimatedLoader from '../svg/AnimatedLoader';

type Props = {}

const Loader: React.FunctionComponent<Props> = (props): JSX.Element => {
  return (
    <div
      css={css`
        justify-content: center;
        text-align: center;
        margin-left: auto;
        margin-right: auto;
      `}
    >
      <AnimatedLoader />
    </div>
  );
};

export default Loader;

```

It compiles. It displays correctly, but in WebStorm, I get a `
TS2322: Type '{ children: Element; css: SerializedStyles; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.   Property 'css' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.`.

![image](https://user-images.githubusercontent.com/3807458/104130167-91f28f00-536f-11eb-8de5-65228c741357.png)

Here is my `babel.config.js`:

```js
/**
 * Babel configuration for Next.js
 * 
 * @see https://nextjs.org/docs/advanced-features/customizing-babel-config Official doc reference v10
 * @see https://github.com/vercel/next.js/blob/canary/packages/next/build/babel/preset.ts You can take a look at this file to learn about the presets included by next/babel.
 * @see https://emotion.sh/docs/css-prop##babel-preset Configuring Emotion 11
 * @example https://github.com/vercel/next.js/tree/canary/examples/with-custom-babel-config Next.js official example of customizing Babel
 */
module.exports = {
  presets: [
    [
      "next/babel",
      {
        "preset-react": {
          "runtime": "automatic",
          "importSource": "@emotion/react"
        }
      }
    ]
  ],
  plugins: ["@emotion/babel-plugin"],
};

```

I'm especially about the `@jsx jsx` stuff. I removed it at some point in v10 and it worked okay, I was using `@emotion/babel-preset-css-prop`. I've removed `@emotion/babel-preset-css-prop` today as it seems it shouldn't be kept anymore. See https://emotion.sh/docs/@emotion/babel-preset-css-prop.

![image](https://user-images.githubusercontent.com/3807458/104130246-e990fa80-536f-11eb-8ecb-9b39f4997884.png)

My understanding is I'm trying to use the "automatic" mode. I followed the doc at https://emotion.sh/docs/css-prop##babel-preset and modified my above-mentioned Babel config.

What am I missing? Automatic mode is enabled, I don't use Pragma mode, I've migrated to the new `@emotion/react`, applied the codemod v11 and ran them. 🤔 

Commits:
- https://github.com/UnlyEd/next-right-now/commit/e90531a97ddcf27605d88dbf74de7e43adc8abc0
- https://github.com/UnlyEd/next-right-now/commit/110662f4ceaf1257e9409cd35f67bee4e0e3ee1f

PR: https://github.com/UnlyEd/next-right-now/pull/241

**Comment by Vadorequest** [comment]
It seems adding `/// <reference types="@emotion/react/types/css-prop" />` to `next-env.d.ts` file fixes the issue. See https://emotion.sh/docs/emotion-11#css-prop-types

Commit: https://github.com/UnlyEd/next-right-now/commit/2a3c89dad00cf212929a54b8d4bb2d14f0677af5

(but I still have other TS issues due to how the `theme` has changed in v11, still on it)

**Comment by Vadorequest** [comment]
Eventually, I had to downgrade to Emotion 10.

The main reason is the babel changes you recommend are far from being compatible with all other vendors, and in my case it broke Cypress entirely. (for other people, it seems it broke MDX)
See https://github.com/vercel/next.js/issues/20952

So, short feedback, but this upgrade was awful from a developer experience standpoint. Not only was it complicated and affected many things (babel, theme, TS types) but in the end I couldn't even benefit from it and had to revert. 😖 

I think you should mention in https://emotion.sh/docs/css-prop#babel-preset using the Babel preset might affect other 3rd party libs (e.g: MDX, Cypress), and using JSX pragma is safer for now.

In my case, it was impossible to know it wouldn't work with Cypress until the very end. I wish other people don't run into this.

**Comment by Andarist** [comment]
The [TypeScript page of our docs](https://emotion.sh/docs/typescript#css-prop) mentions what has to be done to make the css prop work out of the box - TS 4.1 is required for it to work. 

As to next.js - I've successfully used Emotion 11 in next.js project and I didn't really have to configure anything in a special - apart from `next/babel` preset. If this doesn't work for you maybe you are using some older version of next.js?

Not sure what's going with Cypress - would have to get a repro case and clear instructions on how to run a particular project so I could take a look. Cypress is not really responsible directly for anything related to JSX, TS etc - the whole thing is outsourced to webpack (in older versions to browserify). So the only logical explanation for this would be that something is misconfigured but since it's often a delicate interplay between various tools I can't just know what's going on given the provided information.

**Comment by Vadorequest** [comment]
> The TypeScript page of our docs mentions what has to be done to make the css prop work out of the box - TS 4.1 is required for it to work.

Thanks for pointing this out, it's not mentioned in https://emotion.sh/docs/emotion-11#typescript and I had missed it.

I'm using Next.js 10, and the latest version of Cypress.

I made a reproduction at https://github.com/UnlyEd/next-right-now/pull/247 after I reverted the Emotion 11 upgrade, to do it in a separate piece of work. The issue is mentioned at https://github.com/cypress-io/cypress/issues/14471

I'll try using the `jsxImportSource` you mentioned, maybe that's the root cause for there Webpack compilation issue.

**Comment by Vadorequest** [comment]
It seems replacing `cypress/tsconfig.json` `jsx: "react"` option by `"jsxImportSource": "@emotion/react"` fixed the Webpack compilation issue when running Cypress.
See commit [`914996e` (#247)](https://github.com/UnlyEd/next-right-now/pull/247/commits/914996e52505d192055b7733e96371e44f18a1e9)

I'm confused whether this is the right setting or if this should happen in `tsconfig.json` (instead of `cypress/tsconfig.json`). I wonder if part of the issue isn't because the `cypress/tsconfig.json` extends from `tsconfig.json`.

---

I also tried using `"jsxImportSource": "@emotion/react"` in main `tsconfig.json`, and it works too locally (but fails on Vercel). When I do that, I can remove `/// <reference types="@emotion/react/types/css-prop" />` from my `next-env.d.ts` file.
See commit [`a6a0aff` (#247)](https://github.com/UnlyEd/next-right-now/pull/247/commits/a6a0aff0cfb542aaab98cdb721644cce10f61848)

The reason it fails is weird. Out of all my files using `css={css...}` [only one fail](https://github.com/UnlyEd/next-right-now/blob/503e594384fe5f930bda6c7b0f4950d61cc66542/src/components/ComponentTemplate.tsx#L14-L16), using:

```tsx
    <Text
      css={css`
        margin: 30px;
    `}
    >
      {`
        This component is a template meant to be duplicated to quickly get started with new React components.

        Feel free to adapt it at your convenience
      `}
    </Text>
```

I believe the error is mine here, because the [`Text` component](https://github.com/UnlyEd/next-right-now/blob/503e594384fe5f930bda6c7b0f4950d61cc66542/src/components/utils/Text.tsx) doesn't describe a `css` property. I guess its `Props` should extend some native HTML element, but since the said HTML element (`wrapper`) is dynamic, **I'm not sure how it should be described**.

---

So, to summarize, this issue could have been avoided by specifying in https://emotion.sh/docs/emotion-11#typescript we had to use `"jsxImportSource": "@emotion/react"` in the `tsconfig.json`, as specified in https://emotion.sh/docs/typescript#css-prop. That's the requirement to avoid having to add `/// <reference types="@emotion/react/types/css-prop" />` ourselves, and it also fixes my Cypress issue. 🎉

> I encountered an additional bug with the `Text` component, but it's a migration issue from something that used to work. (and which does work when manually importing `/// <reference types="@emotion/react/types/css-prop" />`.

See https://github.com/UnlyEd/next-right-now/pull/247

**Comment by Vadorequest** [comment]
I tried adding Storybook to my project and the `css` has bitten me again.

```
"@storybook/theming@6.1.11":
  version "6.1.11"
  resolved "https://registry.yarnpkg.com/@storybook/theming/-/theming-6.1.11.tgz#0e757e0562aedefc96bd1d12714d4c8e92823467"
  integrity sha512-zRChacVgKoU2BmpvwK1ntiF3KIpc8QblJT7IGiKfP/BNpy9gNeXbLPLk3g/tkHszOvVYtkaZhEXni4Od8tqy1A==
  dependencies:
    "@emotion/core" "^10.1.1"
    "@emotion/is-prop-valid" "^0.8.6"
    "@emotion/styled" "^10.0.23"
    "@storybook/client-logger" "6.1.11"
    core-js "^3.0.1"
    deep-object-diff "^1.1.0"
    emotion-theming "^10.0.19"
    global "^4.3.2"
    memoizerific "^1.11.3"
    polished "^3.4.4"
    resolve-from "^5.0.0"
    ts-dedent "^2.0.0"
```

Because Storybook seem to use Emotion 10, it imports packages that enter in conflict with my Emotion 11 configuration. (I wouldn't have thought about that until I read https://github.com/emotion-js/emotion/issues/971#issuecomment-433567260)

What's the recommandation when importing packages using Emotion 10 while using Emotion 11 ?

**Comment by ljosberinn** [comment]
Over in the Chakra repository we had this issue too and solved it like this: https://github.com/chakra-ui/chakra-ui/blob/master/.storybook/main.js

This is however a Storybook issue, not emotion.

**Comment by Vadorequest** [comment]
Thanks! It seems you're overriding webpack aliases by using webpack aliasing to use Emotion 11 when Emotion 10 is being used. (I'll give a try in a similar way to https://github.com/UnlyEd/next-right-now/blob/f24da10646f1a8225fed537337193fd0707f7ac1/next.config.js#L253)

Although, I don't see why it'd be a Storybook issue, they haven't updated yet to v11, but Storybook shouldn't stop to work because someone upgraded their own Emotion version. I mean, if I had opened an issue there, wouldn't they tell me it's an issue with Emotion itself? 🤔 (I might miss something here)

**Comment by Andarist** [comment]
The shouldnt tell u that this is out issue - its true that your update should not mess things up but it has done it because they use custom webpack aliases. And that's the problem - they shouldnt be doing that in the first place.

As to the IDE problem - have u configured tsconfig with jsxImportSource?

**Comment by DominicTobias-b1** [comment]
I like this idea of declr mergin, however having trouble in CRA.

Firstly declaring an import outside breaks all the other types but works for Emotion in `react-app-env.d.ts` (they just stop working).

```ts
import { Theme as DefaultTheme } from './global/theme'
declare module '@emotion/react' {
  export interface Theme extends DefaultTheme {}
}
```

Found a workaround:

```ts
declare module '@emotion/react' {
  export type Theme = import('./global/theme').Theme
}
```

Then the other types start working and so does the theme, except in both cases @emotion/react looses any of it's other type defs:

![Screenshot 2021-01-22 at 1 36 47 PM](https://user-images.githubusercontent.com/58683337/105451119-e1618500-5cb6-11eb-8072-ab54f5cb6993.png)

🤔

**Comment by giannif** [comment]
As a package maintainer, should I upgrade to Emotion v11? My `theme` is null when using Storybook, because it's v10, causing my package to break in that context.

I'm trying to decide, do I upgrade and roll my own ThemeProvider so I can be sure my package works in all environments? Or leave my package at v10, risking breaking projects that use my package that use Emotion v11?

Any help on the best approach here would be appreciated

**Comment by Andarist** [comment]
The Storybook issue shouldn't be a deciding factor IMHO - it really shouldn't happen and it's a problem with their setup, it's a bug in their code so to speak.

> I'm trying to decide, do I upgrade and roll my own ThemeProvider so I can be sure my package works in all environments? Or leave my package at v10, risking breaking projects that use my package that use Emotion v11?

This is a somewhat hard question to answer right away - I would appreciate it if you could share some more light on how your library is being used. What APIs does it expose to your users and if they are aware of it being Emotion-based - do they primarily use a mixture of your and our APIs or are they limited to your APIs only and they can use a separate Emotion version on their own?

**Comment by giannif** [comment]
@Andarist thanks for the insight. My users have no awareness of Emotion, they render a UI in our [react package](https://github.com/Giphy/giphy-js/blob/master/packages/react-components/README.md) or [vanilla js](https://github.com/Giphy/giphy-js/blob/master/packages/components/README.md) (which uses Preact under the hood), and don't interact with Emotion APIs. 

If only Storybook's setup is the issue here, then I'm not going to worry about my Emotion v10 breaking the host app (which was my biggest concern). I'll probably leave my version at v10 for a bit since I'm on to something else, and upgrade in a few months. Thanks for the help!

**Comment by anilanar** [comment]
Can you even have multiple versions of emotion bundled together? If yes, all good. If not, then @giannif is subject to problems if the host app uses emotion, but another version. The only way to fix that would be to make emotion a peer dependency with a specific version.

**Comment by pleunv** [comment]
Do keep in mind that, when hiding emotion from consumers of your lib, they might still want/need a way to provide a nonce for CSP purposes, or a custom key in order to avoid conflicts if they use emotion internally as well. I think in general it's probably not a bad idea to support users passing in a custom emotion cache.

**Comment by Andarist** [comment]
@pleunv Yep, totally agree - this is why `@emotion/react` has a lazy API that doesn't give you string class names. The `CacheProvider` API is powerful to handle those use cases for consumers without much trouble.

It's best if both sides use the same version of Emotion but things should work if they mix major versions - running 2 instances of let's say Emotion 11 is not supported though. If some scenario doesn't work when mixing 2 major versions then it is a bug in Emotion.
